sound/soc/sof/ipc4-pcm.c

```diff
@@ -281,6 +281,14 @@ static int sof_ipc4_chain_dma_trigger(struct snd_sof_dev *sdev,
 		set_fifo_size = false;
 		break;
 	case SOF_IPC4_PIPE_RESET: /* Disable and free chained DMA. */
+
+		/*
+		 * The ChainDMA can only be reset if it has been initialzied,
+		 * which is only done at stream start.
+		 */
+		if (!spcm->pending_stop[direction])
```

Comment:
Yikes. 

This is re-using a state variable for a completely different purpose.
Are we sure this is the right variable? I already get a migraine to just think of the state machine that will use this.

Also this is only used for pcm_ops->platform_stop_during_hw_free, I don't think you can use it unconditionally.

Reply:
IPC4 always uses this flag on trigger stop, which needs the stream to be started.
I was trying to avoid a new bool[2] for the ChainDMA initialized, but looks like I cannot.
The ChainDMA is a different construct than normal PCMs.

Let me revisit this next week.

Reply:
also is this really an issue that needs to be fixed at the kernel level? Why not fix this at the firmware level? 

Or is this a problem with the reference firmware? I am not sure how we didn't detect this earlier if this was a problem with SOF proper.

Reply:
@plbossart, yes, I have noticed it with the reference firmware (on UP2 after a kernel update).
We did not detect this as we stopped testing the reference firmware with Linux a long time ago and I think the postpone the stop to later in case of IPC4 came later.
This worked some time ago fine. What is not tested is not working (non ChainDMA on HDA with cAVS reference fw also broken).